### ISSUE TITLE
Alt Drag Delay in Waveform (still occur in 3.4.8?)

### ISSUE THREAD

**Issue by LeonCheung** [SubtitleEdit/subtitleedit#1108]
Regarding to my last comments in #238, which has been closed, I think it would be more appropriate to open a new issue since it is still happening, but the cause has been targeted fortunately (I hope).

NOTE:
1. This issue only seems to be more obvious **after working for some minutes**.
2. I believe I got a very high performance PC (24 thread, 32GB RAM, dual GTX680), so I think there should be nothing to do with the harware.

Please see my last comments there for more info, thank you so so so much!

**Comment by niksedk** [comment]
I still cannot re-create this on any of my PCs!
Could you send me your complete settings.xml file?

Can anybody else re-create this?

**Comment by LeonCheung** [comment]
@niksedk [Here](https://www.youtube.com/watch?v=fF62gAHellY) is another video (a bit long, key part is from 15:00), to show from initial setting to working on it and then targeting it. 

BTW,  at 0:15, I cannot customize shortcut for Save operator other than the default setting. Is this a finshy bug?

**Comment by LeonCheung** [comment]
[Here](https://drive.google.com/file/d/0B1ojv5B6Lb-3dUp6dkJSLVpzXzQ/view?usp=sharing) is the current xml file.

**Comment by LeonCheung** [comment]
P.S. I should stress that the problem will disappear after restarting SE, but it will show up again after working for a while.

Now I wonder if there is something to do with the cache or something iterated? Or, it probably about my workflow? (You see how I set S and D shortcut, which are the most commonly way I used to work on subs)

**Comment by niksedk** [comment]
I've fixed a minor memory leak in the waveform (fb1710992ab27d4500b6361347aae3a00b26a638)... could you test if the problem is still there: http://www.nikse.dk/SubtitleEdit.zip (or github source)

**Comment by LeonCheung** [comment]
Hi, First, thank you very much for not forgetting this issue! :P 

After a hard testing on a 30 minutes sub work, I didn't see any obvious fix on this particular one.

Well, there still might be kinda memory leak things existing (since this issue becomes obvious ONLY after working for quite a while), I still believe the key problem appears fairly related to the Syntax coloring rule(s), especially the last checkbox **"Color tme code overlap"**. Because things will go extremely smooth after toggling it off. Would you please take more look into that aspect? I think the suspect was hiding there. :)

Steps to repeat the issue:

1. Use a long-enough sub file, make sure the minimum gap between lines is **1 milisecond**.
2. keep working normally on the subtitle for 5-10 minutes or so (it depends on how frequent your operations are).
3. Then try to adjust in waveform by Alt + dragging (have to test on the active line, since the problem only appears there), to see if you can feel any delay (even a little).
4. If you can feel that (hopefully), then try to toggle off the **Color time code overlap** option, then try to Alt drag again, to see if you can feel better/smoother.

NOTE:

1. It seems to be more obvious with very small gap (e.g. 1 msec) between lines. That's why I suggest to use such sub file in step 1.

2. I got one PC and one laptop, one is Win7 64bit, one is WIn8.1 64 bit, and both have the same issue.

I'll prove it in a screencast video if necessary. (though I've posted more than one video to expose it before)

Thanks again.

**Comment by jdpurcell** [comment]
I wonder if it has something to do with the duration column changing colors. When the color changes, I noticed the listview flickers.  Maybe that would explain why it only happens when resizing the selected subtitle (because the duration validation only happens when a subtitle is selected).

**Comment by niksedk** [comment]
@jdpurcell: Probably, but I cannot re-create it... I've looked for a memory leak bit did not see it in the syntax coloring code (found a smaller one in the waveform code which was fixed a little while back).

@LeonCheung: It would be interesting to see you video with a Task Manager visible...

**Comment by LeonCheung** [comment]
@niksedk I've just upload a video **[here](https://youtu.be/5kH7HCJZyuU)** based on a compile version after @jdpurcell 's latest commit.

Perhaps my workflow is a bit different. You can try to follow the setting I did at the first 30 secs in the video, to see if you can repeat it by then.

I want to stress one thing: Things will become normal right after restarting SE, but will go over and over again, you know...

I'm not a coder, but I'm thinking about something:

When I was using the **Set end, offset the rest and go to next** shortcut, all timecodes after the active subtitle are changed together, I mean, perhaps when something are kinda iterated and need more and more calculations when I did it time and time again? (in my setting, I mean keep hitting D key while editing), and finally lead to some conflict to the **syntax coloring rules**, especially the **Color time code overlap**?

**Comment by LeonCheung** [comment]
BTW, around 01:00 in that video you can see the sections are gone in waveform window for a whle, I never experienced that perviously. I assume there are something fishy in lastest version?

**Comment by jdpurcell** [comment]
Thanks to your new video and the tip about "set end, offset the rest...", I can reproduce this now.  So far I've determined that a large amount of CPU time is being spent in the UpdateOverlapErrors method.  I think the problem is not necessarily that method itself, but that it gets called more and more frequently after more edits are made.  I'm still working on tracking down why that's happening.

**Comment by LeonCheung** [comment]
Oh you are my hero, @jdpurcell !!! Just tested, it works perfect now! You know, you just solved a problem that made me pulling my hairs for a year!!
I don't know how to show my appreciation. May god bless you!
And big thanks to @niksedk too. I think that is exactly the magic of opensource.  :bulb:

**Comment by jdpurcell** [comment]
You're welcome, I'm glad it's working now!

I'm sure Nikolaj would have fixed it too, I just got to it first :).  The reason why it's finally fixed is that you figured out how to quickly and reliably make the problem occur.  Without being able to reproduce it, it would have been nearly impossible to find the problem in the code.

**Comment by niksedk** [comment]
Nice :)
What was `numericUpDownDuration.ValueChanged += NumericUpDownDurationValueChanged;` doing there ? ;)
